src/main/java/com/mattworzala/canary/server/givemeahome/StructureLoader.java

```diff
@@ -0,0 +1,145 @@
+package com.mattworzala.canary.server.givemeahome;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import net.minestom.server.MinecraftServer;
+import net.minestom.server.command.builder.arguments.minecraft.ArgumentBlockState;
+import net.minestom.server.instance.block.Block;
+import net.minestom.server.instance.block.BlockHandler;
+import org.jglrxavpok.hephaistos.nbt.NBTCompound;
+import org.jglrxavpok.hephaistos.nbt.NBTException;
+import org.jglrxavpok.hephaistos.nbt.SNBTParser;
+
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.List;
+
+public class StructureLoader {
+
+    public static Block parseBlock(JsonElement block) {
+        String blockField;
+        String handlerField = "";
+        String dataField = "";
+
+        if (block.isJsonObject()) {
+            JsonObject blockObj = block.getAsJsonObject();
+
+            blockField = blockObj.get("block").getAsString();
+
+            if (blockObj.get("handler") != null) {
+                handlerField = blockObj.get("handler").getAsString();
+            }
+            if (blockObj.get("data") != null) {
+                dataField = blockObj.get("data").getAsString();
+            }
+        } else {
+            blockField = block.getAsString();
+        }
+        var argBlockState = new ArgumentBlockState("blockStateId");
+        Block b = argBlockState.parse(blockField);
+
+        if (handlerField.length() > 0) {
+            final BlockHandler handler = MinecraftServer.getBlockManager().getHandler(handlerField);
+            b = b.withHandler(handler);
+        }
+        if (dataField.length() > 0) {
+            SNBTParser parser = new SNBTParser(new StringReader(dataField));
+            try {
+                b = b.withNbt((NBTCompound) parser.parse());
+            } catch (NBTException e) {
+                e.printStackTrace();
+            }
+        }
+        return b;
+    }
+
+    public static List<Integer> parseSize(JsonArray sizeArr) {
+        var sizeList = new ArrayList<Integer>();
+        for (JsonElement elem : sizeArr) {
+            sizeList.add(elem.getAsInt());
+        }
+        return sizeList;
+    }
+
+    public static List<Structure.BlockDef> parseBlocks(String blocks) throws Exception {
+        List<Structure.BlockDef> list = new ArrayList<>();
+        for (String s : blocks.split(";")) {
+            Structure.BlockDef blockDef = parseBlockDef(s);
+            list.add(blockDef);
+        }
+        return list;
+    }
+
+    public static Structure.BlockDef parseBlockDef(String blockDef) throws Exception {
```

Comment:
no need to be public, also see my note on the writer.

goes for all of these `parseXXX` methods.